### PR TITLE
Embark Anyone

### PR DESCRIPTION
**Pull request by ProtonNumber**
Allows you to embark as any civilization, including dead and non-dwarven ones through a unretire-anyone style UI.

Currently suffers from some issues with counting sites, especially with non-dwarven civs. This is ultimately a cosmetic issue, but I'd be interested if anyone has a more accurate method of doing this.

### CODE REVIEW

File: docs/embark-anyone.rst [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+
+::
+
+    embark-anyone
+
+Note
+-----
+Non-dwarven civs have their own mechanics which can render fortress mode difficult
+to unplayable. Preparing carefully is advised, and cheating in some items may be
+required.
```
[the remark this answers is not in the document]
Comment (ab9rf) [reply docs/embark-anyone.rst:19]: i'd avoid the pejorative "cheating" entirely and instead say something like "without spawning them in manually" or some other phrase that describes the action neutrally

don't embed judgments about how people choose to play the game into the documentation

[the remark this answers is not in the document]
Comment (ProtonNumber) [reply docs/embark-anyone.rst:19]: How about "some crucial items may need to be spawned in"?

File: docs/embark-anyone.rst [hunk 2]
```diff
@@ -0,0 +1,18 @@
+embark-anyone
+=============
+
+.. dfhack-tool::
+    :summary: Allows you to embark as any civilisation, including dead and non-dwarven ones.
+    :tags: embark gameplay armok
```
Comment (myk002) [inline docs/embark-anyone.rst:6]: TODO: think about the implications of combining `gameplay` and `armok`

(If I submit this review without finishing my thoughts here, please poke me about this)

File: docs/embark-anyone.rst [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+
```
Comment (myk002) [inline docs/embark-anyone.rst:7]: Please add an introductory paragraph before the Usage section that at least goes over when to run the command

File: embark-anyone.lua [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+function embarkAnyone()
+   local viewscreen = dfhack.gui.getDFViewscreen(true)
+
+
+   if viewscreen._type ~= df.viewscreen_choose_start_sitest then
+      qerror("This script can only be used on the embark screen!")
+   end
+
+   for i, civ in ipairs (df.global.world.entities.all) do
+      if civ.type == df.historical_entity_type.Civilization then
```
Comment (myk002) [inline embark-anyone.lua:22]: instead of placing the entire body of the for loop in an if statement, consider inverting the logic and skipping the for body (that is, a "continue" statement in other languages) using a Lua label:
```lua
for i, civ in ipairs(df.global.world.entities.all) do
    if civ.type ~= df.historical_entity_type.Civilization then goto continue end
    ...
    ::continue::
end
```
This reduces the overall indentation and can make the code easier to reason about and follow

File: embark-anyone.lua [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+         local available = false
+
+         -- Check if civ is already available to embark
+         for _, item in ipairs (viewscreen.start_civ) do
+            if item == civ then
+               available = true
+               break
+            end
+         end
```
Comment (myk002) [inline embark-anyone.lua:31]: this would be more efficient if you built a set outside of this loop. e.g.:
```lua
local existing_civs = {}
for _,existing_civ in ipairs(viewscreen.start_civ) do
    existing_civs[existing_civ.id] = true
end
```
Then later you can test for membership by evaluating `existing_civs[civ.id]`:
```lua
if existing_civs[civ.id] then goto continue end
```

File: embark-anyone.lua [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+                  -- DF stores population info as an array of groups of residents (?).
+                  -- Inspecting these further could give a more accurate count.
+                  for _, group in ipairs(site.populace.inhabitants) do
+                     pops = pops + group.count
+                  end
+               end
+
+               -- Count living nemeses
+               for _, nem in ipairs (civ.nemesis_ids) do
+                  if df.global.world.nemesis.all[nem].figure.died_year == -1 then
```
Comment (myk002) [inline embark-anyone.lua:56]: nemesis IDs are *usually* sequential, but not always. Use `df.nemesis.find(nem)` instead

File: embark-anyone.lua [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+            local sites = 0
+            local pops = 0
+            local nemeses = 0
+            local histfigs = 0
+            local label = ''
+
+            -- Civs keep links to sites they no longer hold, so check owner
+            -- We also take the opportunity to count population
+            for j, link in ipairs(civ.site_links) do
+               local site = df.global.world.world_data.sites[link.target]
```
Comment (myk002) [inline embark-anyone.lua:43]: if `link.target` is the site id, you'd get the site with `df.site.find(link.target)`, which does a binary search through the sites vector. the ID is not necessarily the same as the offset.

File: embark-anyone.lua [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+               for _, nem in ipairs (civ.nemesis_ids) do
+                  if df.global.world.nemesis.all[nem].figure.died_year == -1 then
+                     nemeses = nemeses + 1
+                  end
+               end
+
+               -- Count living histfigs
+               -- Used for death detection. May be redundant.
+               for _, fig in ipairs (civ.histfig_ids) do
+                  if df.global.world.history.figures[fig].died_year == -1 then
```
Comment (myk002) [inline embark-anyone.lua:64]: here too. use `df.historical_figure.find(fig)`. I didn't mention this above, but you also have to test the result of `find` for nil, since it's possible that the IDs in the vectors are stale, and the referent has been deleted.

File: embark-anyone.lua [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+            label = label .. dfhack.units.getRaceNamePluralById(civ.race) .. "\n"
+
+            -- Add pop & site count or mark civ as dead.
+            if histfigs == 0 and pops == 0 then
+               label = label .. "Dead"
+            else
+               label = label .. "Pop: " .. (pops + nemeses) .. " Sites: " .. sites
+            end
+
+            table.insert(choices, {text = label, search_key = label:lower(),
```
Comment (myk002) [inline embark-anyone.lua:89]: you can let the search key be automatically generated if it's just going to be a version of the label (the match is case insensitive anyway)

File: embark-anyone.lua [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+                                   info = {civ = civ, pops = pops, sites = sites,
+                                           nemeses = nemeses}})
+         end
+      end
+   end
+   dialogs.ListBox{
```
Comment (myk002) [inline embark-anyone.lua:95]: you should probably handle the case where there are no more civs to add to the list. A dialogs.MessageBox with the message should be sufficient.

File: docs/embark-anyone.rst (hunk 2 above)
Comment (myk002) [reply docs/embark-anyone.rst:6]: more words:

I've been treating "gameplay" and "armok" as mutually exclusive, with "gameplay" meaning "changes the gameplay in some consistent way that may not be achievable in vanilla but is not inherently "cheaty" or game-breaking" and armok meaning "no restrictions -- the world conforms to your will".

In other words, "gameplay" tools can have some "armoky" elements to them, but they are intended to be a balanced extension to the world.

This script straddles that line closer than most other tools.

Both `unretire-anyone` and `gui/embark-anywhere` have the `armok` tag and not the `gameplay` tag, so my suspicion is that we should go with `:tags: embark armok` here and remove the `gameplay` tag.

Comment (ProtonNumber) [reply docs/embark-anyone.rst:6]: Given that some civs are borderline unplayable, I'll agree that armok alone is appropriate.

File: embark-anyone.lua (hunk 4 above)
Comment (ProtonNumber) [reply embark-anyone.lua:22]: Ahhhh okay

Sorry, I'm still used to C, where using goto is _heavily_ frowned upon.

Comment (myk002) [reply embark-anyone.lua]: totally understood. I had to get used to this Lua-ism as well

File: docs/embark-anyone.rst [hunk 11]
```diff
@@ -0,0 +1,23 @@
+embark-anyone
+=============
+
+.. dfhack-tool::
+    :summary: Allows you to embark as any civilisation, including dead and non-dwarven ones.
+    :tags: embark armok
+
+This script can be ran on the embark screen and presents a list of civilisations that
+are not normally playable to add to the "Select Origin Civ" menu. This includes dead
+civilisations, the normal non-dwarven civilisations, and subterranean animal people.
```
Comment (myk002) [inline docs/embark-anyone.rst:10]:
```suggestion
This script must be run on the embark screen when choosing an origin civilization.
When run, you can add any civilization, including dead or non-dwarven civilizations,
to the list of choices.
```

File: docs/embark-anyone.rst [hunk 12]
```diff
@@ -0,0 +1,23 @@
...
+-----
+
+::
+
+    embark-anyone
+
+Note
+-----
+Non-dwarven civs have their own mechanics which can render fortress mode difficult
+to unplayable. Preparing carefully is advised, and some crucial items may need to be spawned in.
```
Comment (myk002) [inline docs/embark-anyone.rst:23]:
```suggestion
or unplayable. Preparing carefully is advised, and some crucial items may need to be
spawned in with other DFHack tools (e.g. `gui/create-item`).
```

File: embark-anyone.lua [hunk 13]
```diff
@@ -0,0 +1,117 @@
...
+
+   viewscreen.start_civ:insert ('#', info.civ)
+   viewscreen.start_civ_nem_num:insert ('#', info.nemeses)
+   viewscreen.start_civ_entpop_num:insert ('#', info.pops)
+   viewscreen.start_civ_site_num:insert ('#', info.sites)
+end
+
+function embarkAnyone()
+   local viewscreen = dfhack.gui.getDFViewscreen(true)
+   local existing_civs = {}
```
Comment (myk002) [inline embark-anyone.lua:16]:
```suggestion
   local choices, existing_civs = {}, {}
```

File: embark-anyone.lua [hunk 14]
```diff
@@ -0,0 +1,117 @@
+local dialogs = require 'gui.dialogs'
+local utils = require 'utils'
+local choices = {}
```
Comment (myk002) [inline embark-anyone.lua:3]:
```suggestion
local dialogs = require('gui.dialogs')
local utils = require('utils')
```

File: embark-anyone.lua [hunk 15]
```diff
@@ -0,0 +1,117 @@
...
+      label = label .. dfhack.units.getRaceNamePluralById(civ.race) .. "\n"
+
+      -- Add pop & site count or mark civ as dead.
+      if histfigs == 0 and pops == 0 then
+         label = label .. "Dead"
+      else
+         label = label .. "Pop: " .. (pops + nemeses) .. " Sites: " .. sites
+      end
+
+      table.insert(choices, {text = label, search_key = label:lower(),
```
Comment (myk002) [inline embark-anyone.lua:89]:
```suggestion
      table.insert(choices, {text = label,
```

File: embark-anyone.lua [hunk 16]
```diff
@@ -0,0 +1,117 @@
...
+                             info = {civ = civ, pops = pops, sites = sites,
+                                     nemeses = nemeses}})
+
+      ::continue::
+   end
+
+   if #choices then
```
Comment (myk002) [inline embark-anyone.lua:96]:
```suggestion
   if #choices > 0 then
```

Comment (myk002) [reply embark-anyone.lua:96]: in Lua, 0 is truthy

File: embark-anyone.lua [hunk 17]
```diff
@@ -0,0 +1,117 @@
...
+      dialogs.ListBox{
+         frame_title = 'Embark Anyone',
+         text = 'Select a civilisation to add to the list of origin civs:',
```
Comment (myk002) [inline embark-anyone.lua:99]:
```suggestion
         text = 'Select a civilization to add to the list of origin civs:',
```

File: embark-anyone.lua [hunk 18]
```diff
@@ -0,0 +1,117 @@
...
+         on_select = function(id, choice)
+            addCivToEmbarkList(choice.info)
+         end,
+         with_filter = true,
+         row_height = 4,
+      }:show()
+   else
+      dialogs.MessageBox{
+         frame_title = 'Embark Anyone',
+         text = 'No additional civilisations found.'
```
Comment (myk002) [inline embark-anyone.lua:111]:
```suggestion
         text = 'No additional civilizations found.'
```

Comment (myk002) [reply embark-anyone.lua:111]: using american spelling to conform to other uses in the game and on the wiki

File: embark-anyone.lua (hunk 14 above)
Comment (myk002) [reply embark-anyone.lua:3]: use preferred formatting; move `choices` to block where it is used

File: embark-anyone.lua (hunk 15 above)
Comment (myk002) [reply embark-anyone.lua:89]: not only is the search key correctly autogenerated in this case, the autogenerated version additionally handles special characters in names (e.g. umlauts and other diacritics)

**Comment by myk002** [comment]
At your convenience, please also add your name to https://github.com/DFHack/dfhack/blob/develop/docs/about/Authors.rst

**Review by myk002 — COMMENTED** [review]
looks great! there's one last thing I forgot to mention: could you add a line under the "New tools" header in the "Future" version in `changelog.txt`?

https://github.com/DFHack/scripts/blob/master/changelog.txt#L30

**Review by myk002 — APPROVED** [review]
